api/v1alpha1/gateway_types.go

```diff
@@ -179,56 +182,95 @@ type ListenerTLS struct {
 
 // GatewayStatus defines the observed state of Gateway
 type GatewayStatus struct {
-	// TODO overall status
-
-	// Listeners is the status for each listener block in the
-	// Spec. The status for a given block will match the order as
-	// declared in the Spec, e.g. the status for Spec.Listeners[3]
-	// will be in Status.Listeners[3].
+	// Conditions describe the current conditions of the Gateway.
+	Conditions []GatewayCondition `json:"conditions"`
+	// Listeners provide status for each listener defined in the Spec. The name
+	// in ListenerStatus refers to the corresponding Listener of the same name.
 	Listeners []ListenerStatus `json:"listeners"`
-	// Routes is the status for each attached route to the
-	// Gateway. The status for a given route will match the orer as
-	// declared in the Spec, e.g. the status for Spec.Routes[3] will
-	// be in Status.Routes[3].
-	Routes []GatewayRouteStatus `json:"routes"`
+}
+
+// GatewayConditionType is a type of condition associated with a Gateway.
+type GatewayConditionType string
+
+const (
+	// ConditionNoSuchGatewayClass indicates that the specified GatewayClass
+	// does not exist.
+	ConditionNoSuchGatewayClass GatewayConditionType = "NoSuchGatewayClass"
+	// ConditionGatewayNotScheduled indicates that the Gateway has not been
+	// scheduled.
+	ConditionGatewayNotScheduled GatewayConditionType = "GatewayNotScheduled"
+	// ConditionListenersNotReady indicates that at least one of the specified
+	// listeners os not ready.
+	ConditionListenersNotReady GatewayConditionType = "ListenersNotReady"
+	// ConditionInvalidListeners indicates that at least one of the specified
+	// routes is invalid.
+	ConditionInvalidListeners GatewayConditionType = "InvalidListeners"
+	// ConditionRoutesNotReady indicates that at least one of the specified
+	// routes is not ready.
+	ConditionRoutesNotReady GatewayConditionType = "RoutesNotReady"
+	// ConditionInvalidRoutes indicates that at least one of the specified
+	// routes is invalid.
+	ConditionInvalidRoutes GatewayConditionType = "InvalidRoutes"
```

Comment:
How does the operator/tooling know which route was invalid? There probably should be some way to link this to the broken route that is more deterministic than scanning the `Message`. Maybe the same approach as with Listeners (specific route conditions in `.status.routes.conditions`) could work.

I think that it is fine to leave a TODO and defer this to a separate issue.

Reply:
Yeah routes are tough to get right here. Theoretically all status for an individual route could exist on that API resource. The potential quantity of routes that could be contained within a Gateway makes me hesitant to add status for each one within the Gateway resource. It seems like we'd be at risk of hitting object size limits here. I think how we handle route status here should be dependent on what we determine is a reasonable limit for the number of routes that should be contained/referenced within a Gateway. I do agree that it would be nice to have status for all the things in one place, so if we can accomplish that, it would be good. Like you suggested, this may be better for a TODO and/or separate issue.